### PR TITLE
Allow getting the share list filtered by share type via API

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Allow filtering by share type while getting the shares. This should speed up a bit the web UI because the server will need to fetch and send less data.
Changes are backwards-compatible: if no filter is sent, all the shares will be returned, same as before.

## Related Issue
https://github.com/owncloud/enterprise/issues/4107

## Motivation and Context
The "share by link" view is expected to perform faster since there are less requests done to the DB to fetch data. Furthermore, the view doesn't need to ask for remote shares, which could slow down things if there are problems with the remote server.

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by jvillafanez** [comment]
Missing items:
* [ ] Add new unit tests for the new cases
* [ ] Add changelog entry

Code has been checked via web UI and partially via API

**Comment by phil-davis** [comment]
Could also add a couple of API acceptance tests. There is a step:
`When user "Alice" gets all shares shared by him using the sharing API`

That could have more specific steps to get just the user, group or public link shares.

The scenario(s) could create a few shares of different types, then do a "get" filtering by share type.

**Comment by micbar** [comment]
@jvillafanez Let us include this in 10.6

**Comment by phil-davis** [comment]
@jvillafanez there are some JavaScript tests that failed: https://drone.owncloud.com/owncloud/core/27205/8/6
Are those real fails that need to be fixed?

**Comment by jvillafanez** [comment]
Failing test fixed, and new unit tests added. This should be ready now.

**Comment by phil-davis** [comment]
As user admin, I shared a folder `f1` with user `phil` and folder `f2` with group `grp1`. `phil` id a member of `grp1`

When I query as `phil` to get the group shares that are shared with me, I get both shares returned:
```
curl -u phil:phil http://172.17.0.1:8080/ocs/v2.php/apps/files_sharing/api/v1/shares?share_types=1\&shared_with_me=true
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>200</statuscode>
  <message/>
  <totalitems></totalitems>
  <itemsperpage></itemsperpage>
 </meta>
 <data>
  <element>
   <id>344</id>
   <share_type>0</share_type>
   <uid_owner>admin</uid_owner>
   <displayname_owner>I am admin</displayname_owner>
   <permissions>31</permissions>
   <stime>1603167789</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>admin</uid_file_owner>
   <displayname_file_owner>I am admin</displayname_file_owner>
   <additional_info_owner/>
   <additional_info_file_owner/>
   <state>0</state>
   <path>/f1</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>shared::/f1</storage_id>
   <storage>739</storage>
   <item_source>2147525538</item_source>
   <file_source>2147525538</file_source>
   <file_parent>2147525335</file_parent>
   <file_target>/f1</file_target>
   <share_with>phil</share_with>
   <share_with_displayname>Phil Ji</share_with_displayname>
   <share_with_additional_info/>
   <mail_send>0</mail_send>
   <attributes/>
  </element>
  <element>
   <id>499</id>
   <share_type>1</share_type>
   <uid_owner>admin</uid_owner>
   <displayname_owner>I am admin</displayname_owner>
   <permissions>31</permissions>
   <stime>1603175000</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>admin</uid_file_owner>
   <displayname_file_owner>I am admin</displayname_file_owner>
   <additional_info_owner/>
   <additional_info_file_owner/>
   <state>0</state>
   <path>/f2</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>shared::/f2</storage_id>
   <storage>739</storage>
   <item_source>2147525539</item_source>
   <file_source>2147525539</file_source>
   <file_parent>2147525335</file_parent>
   <file_target>/f2</file_target>
   <share_with>z111</share_with>
   <share_with_displayname>z111</share_with_displayname>
   <mail_send>0</mail_send>
   <attributes/>
  </element>
 </data>
</ocs>
```
The filtering did not happen.

But when I query as `admin` for the group shares that have been shared by me, I correctly get just the group share:
```
curl -u admin:admin http://172.17.0.1:8080/ocs/v2.php/apps/files_sharing/api/v1/shares?share_types=1
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>200</statuscode>
  <message/>
  <totalitems></totalitems>
  <itemsperpage></itemsperpage>
 </meta>
 <data>
  <element>
   <id>499</id>
   <share_type>1</share_type>
   <uid_owner>admin</uid_owner>
   <displayname_owner>I am admin</displayname_owner>
   <permissions>31</permissions>
   <stime>1603175000</stime>
   <parent/>
   <expiration/>
   <token/>
   <uid_file_owner>admin</uid_file_owner>
   <displayname_file_owner>I am admin</displayname_file_owner>
   <additional_info_owner/>
   <additional_info_file_owner/>
   <path>/f2</path>
   <item_type>folder</item_type>
   <mimetype>httpd/unix-directory</mimetype>
   <storage_id>home::admin</storage_id>
   <storage>739</storage>
   <item_source>2147525539</item_source>
   <file_source>2147525539</file_source>
   <file_parent>2147525254</file_parent>
   <file_target>/f2</file_target>
   <share_with>z111</share_with>
   <share_with_displayname>z111</share_with_displayname>
   <mail_send>0</mail_send>
   <attributes/>
  </element>
 </data>
</ocs>
```

Is this filtering-by-share-type supposed to work for `shared_with_me` ?

**Comment by phil-davis** [comment]
ToDo: add some API acceptance tests for filtering the `shared_with_me` list. @phil-davis this in progress.

**Comment by phil-davis** [comment]
How will clients know that the server is capable of providing a filtered list of shares by share type?
Should this capability be advertised in the capabilities endpoint?

**Comment by jvillafanez** [comment]
I'd vote for API docs.
For the clients, I think the easiest options are:
* Ignore this feature and keep requesting all the shares. This means they won't need to change anything on their end.
* Request specific share types but keep filtering locally. Local filtering will still be required because the servers prior 10.6 will ignore the share type filtering and return all the share types

I think those options are the ones that require the least changes both in the client and the server.

**Comment by jvillafanez** [comment]
Doc ticket opened in https://github.com/owncloud/docs/issues/2825

**Review by phil-davis — COMMENTED** [review]
API acceptance tests are done for the interesting combinations - looks good.

File: apps/files_sharing/lib/Controller/Share20OcsController.php [hunk 1]
```diff
@@ -655,6 +689,31 @@ public function getShares() {
...
+		} else {
+			$shareTypes = \explode(',', $shareTypes);
+		}
+
+		$requestedShareTypes = [
+			Share::SHARE_TYPE_USER => false,
+			Share::SHARE_TYPE_GROUP => false,
+			Share::SHARE_TYPE_LINK => false,
+			Share::SHARE_TYPE_REMOTE => false,
+		];
```
Comment (karakayasemi) [inline apps/files_sharing/lib/Controller/Share20OcsController.php:709]: IMO, we can simplify the adding share foreach logic with a control like the one below. 
```suggestion
		];
		if ($this->shareManager->outgoingServer2ServerSharesAllowed() === false) {
			\unset($requestedShareTypes[Share::SHARE_TYPE_REMOTE])
		}
```

**Review by karakayasemi — COMMENTED** [review]
Generally, code looks okay. All the changes are in controller and all tests are passed. So, affected parts should be limited. 
I just only suggested a small change that can simplify some foreach blocks. @jvillafanez I would like to hear your thoughts about that.

File: apps/files_sharing/lib/Controller/Share20OcsController.php (hunk 1 above)
Comment (jvillafanez) [reply apps/files_sharing/lib/Controller/Share20OcsController.php:705]: code adjusted

File: apps/files_sharing/lib/Controller/Share20OcsController.php [hunk 2]
```diff
@@ -655,6 +685,36 @@ public function getShares() {
...
+
+		if ($this->shareManager->outgoingServer2ServerSharesAllowed() === false) {
+			// if outgoing remote shares aren't allowed, the remote share type can't be chosen
+			unset($requestedShareTypes[Share::SHARE_TYPE_REMOTE])
```
Comment (phil-davis) [inline apps/files_sharing/lib/Controller/Share20OcsController.php:709]:
```suggestion
			unset($requestedShareTypes[Share::SHARE_TYPE_REMOTE]);
```

**Review by phil-davis — COMMENTED** [review]
Syntax error needs fixing - @jvillafanez

File: apps/files_sharing/lib/Controller/Share20OcsController.php (hunk 2 above)
Comment (jvillafanez) [reply apps/files_sharing/lib/Controller/Share20OcsController.php]: that happens when you overconfidently commit a change without checking... :eyes: 
Fixed now

**Review by karakayasemi — APPROVED** [review]
LGTM!

**Pull request merged**